src/resolution/resolver.ts

```diff
@@ -51,103 +31,86 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
         });
 
     } else {
-
-        let result: undefined | T | Promise<T> | (T | Promise<T>)[];
-
         if (request.target.isOptional() && bindings.length === 0) {
             return undefined;
         }
 
         const binding = bindings[0];
-        const isSingleton = binding.scope === BindingScopeEnum.Singleton;
-        const isRequestSingleton = binding.scope === BindingScopeEnum.Request;
 
-        if (isSingleton && binding.activated) {
-            return binding.cache;
-        }
+        return _resolveBinding<T>(requestScope, request, binding);
+    }
+};
 
-        if (
-            isRequestSingleton &&
-            requestScope !== null &&
-            requestScope.has(binding.id)
-        ) {
-            return requestScope.get(binding.id);
-        }
+const _resolveFactoryFromBinding = <T>(
+    binding:interfaces.Binding<T>,
+    context:interfaces.Context
+): T | Promise<T> => {
+    const factoryDetails = _getFactoryDetails(binding) as FactoryDetails;
+    return _tryAndThrowErrorIfStackOverflow(
+        () => (factoryDetails.factory as FactoryTypeFunction).bind(binding)(context),
+        () => new Error(
+        ERROR_MSGS.CIRCULAR_DEPENDENCY_IN_FACTORY(factoryDetails.factoryType, context.currentRequest.serviceIdentifier.toString())
+    ));
+}
 
-        if (binding.type === BindingTypeEnum.ConstantValue) {
-            result = binding.cache;
-            binding.activated = true;
-        } else if (binding.type === BindingTypeEnum.Function) {
-            result = binding.cache;
+const _getResolvedFromBinding = <T>(
+    requestScope: interfaces.RequestScope,
+    request: interfaces.Request,
+    binding:interfaces.Binding<T>): T | Promise<T> => {
+    let result: T | Promise<T> | undefined;
+    const childRequests = request.childRequests;
+    _ensureFullyBound(binding);
+    switch(binding.type){
+        case "ConstantValue":
+        case "Function":
+            result = binding.cache as T | Promise<T>;
```

Comment:
I don't think that this should be here.
`binding.activated = true;`
This is related to singleton scope not a resolved value from the binding type.  After it is resolved from binding the singleton scope sets binding.activated to true.

Reply:
We can do it in this PR, I created an issue to solve this so I think it's ok to change it

Reply:
Great !

Reply:
Ok for you to change these ?

Reply:
One other thing that could be changed here given we are refactoring "resolve"

NextArgs can be generic as well as GetArgs.
Then _getAllArgs / _getNotAllArgs and _planAndResolve can be changed.

Reply:
Nice! I removed `bindingActivated` assignment. Can you update `NextArgs` and `GetArgs`?

Reply:
Will do be about 30 minutes